sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,760 @@
+with the allocation kind `usm::alloc::host`.
+
+[source, c++]
+----
+memory_pool context::ext_oneapi_get_default_memory_pool(device dev,
+  usm::alloc kind) const;
+----
+
+_Returns_: The default memory pool associated with the context and `dev` for
+allocating with the allocations of kind `kind`.
+
+_Throws_: An exception with the `errc::invalid` error code if `kind` is
+`usm::alloc::host`.
+
+
+=== Asynchronous malloc & free
+
+This extension introduces a series of new enqueue functions for enqueueing
+asynchronous malloc and free commands which operate with the memory pools also
+introduced in this extension.
+
+All enqueue functions introduced have overloads which take a SYCL `queue` and a
+SYCL `handler`. None of enqueue functions return a SYCL `event` directly, as
+this extension is in line with the
+link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[
+  sycl_ext_oneapi_enqueue_functions] extension, so events are returned when
+calling `submit_with_event` and the `handler` overloads of these enqueue
+functions.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental {
+
+void *async_malloc(queue q, usm::alloc kind, size_t size);
+
+void *async_malloc(handler h, usm::alloc kind, size_t size);
+
+void *async_malloc_from_pool(queue q, size_t size, memory_pool pool);
+
+void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);
+
+void async_free(queue q, void *ptr);
+
+void async_free(handler h, void *ptr);
+
+}  // ext::oneapi::experimental
+----
+
+[source, c++]
+----
+void *async_malloc(queue q, usm::alloc kind, size_t size);
+
+void *async_malloc(handler h, usm::alloc kind, size_t size);
+
+void *async_malloc_from_pool(queue q, size_t size, memory_pool pool);
+
+void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);
+----
+
+_Effects_: Enqueues a command to `q` or the SYCL queue associated with `h` which
+will asynchronously allocate memory of size `size` in bytes, allocating from the
+memory pool `pool` if provided, otherwise allocation from the default memory
+pool associated with the SYCL context and device associated with `q` or `h`.
+If a memory pool `pool` is provided the allocation kind is that of the memory
+pool, otherwise if an allocation kind `kind` parameter is provided, this is the
+allocation kind which is used. If a memory pool `pool` is provided, this must be
+associated with the same SYCL context and device as `q` or the SYCL queue
+associated with `h`. Memory is first allocated from the memory pool if possible,
+otherwise memory is allocated from the source to the memory pool to provide
+enough memory in the memory pool for the allocation. Accessing the memory at the
+address of the pointer returned by asynchronous malloc functions before the
+command has completed execution is undefined behavior.
+
+_Returns_: A pointer to the address of a memory reservation if `size` is
+non-zero, otherwise returns `nullptr`.
```

Comment:
This does not match match the core SYCL behavior of the USM allocation functions.  The core spec says:

> When the allocation size is zero bytes (numBytes or count is zero), these functions behave in a manner consistent with C++ std::malloc. The value returned is unspecified in this case, and the returned pointer may not be used to access storage. If this pointer is not null, it must be passed to sycl::free to avoid a memory leak.